File: go/read_narrow_result_set_benchmark.go
package main

import (
	"context"
	"time"

	"cloud.google.com/go/spanner"
	"go.opentelemetry.io/otel/metric"
	"google.golang.org/api/iterator"
)

type ReadNarrowResultSetBenchmark struct {
	numRows              int64
	readLatencyHistogram metric.Float64Histogram
	statement            spanner.Statement
}

func NewReadNarrowResultSetBenchmark(readLatencyHistogram metric.Float64Histogram, numRows int64) *ReadNarrowResultSetBenchmark {
	sql := `SELECT
  FARM_FINGERPRINT(GENERATE_UUID()) AS random_int64_1,
  FARM_FINGERPRINT(GENERATE_UUID()) AS random_int64_2
FROM UNNEST(GENERATE_ARRAY(1, @num_rows)) AS n`

	return &ReadNarrowResultSetBenchmark{
		numRows:              numRows,
		readLatencyHistogram: readLatencyHistogram,
		statement: spanner.Statement{
			SQL:    sql,
			Params: map[string]interface{}{"num_rows": numRows},
		},
	}
}

func (b *ReadNarrowResultSetBenchmark) Name() string {
	return "Read Narrow Result Set Benchmark"
}

func (b *ReadNarrowResultSetBenchmark) Type() string {
	return "read-narrow-result-set"
}

// INTENTIONAL: Do not change ShouldMeasureEntireMethod to return true.
// We intentionally exclude the initial query execution and the first row fetch
// to measure purely the iteration and decoding latency of the remaining rows.
func (b *ReadNarrowResultSetBenchmark) ShouldMeasureEntireMethod() bool {
	return false
}

func (b *ReadNarrowResultSetBenchmark) Execute(ctx context.Context, client *spanner.Client, tableName string, minId, maxId int64) error {
	iter := client.Single().Query(ctx, b.statement)
	defer iter.Stop()

	row, err := iter.Next()
	if err == iterator.Done {
		return nil
	}
	if err != nil {
		return err
	}

	// Decode first row fully to native types
	b.decodeRow(row)

	// Start timing iteration of remaining rows
	start := time.Now()
	for {
		row, err = iter.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return err
		}
		b.decodeRow(row)
	}
	durationUs := float64(time.Since(start).Microseconds())

	// Retrieve OTel attributes from context and record custom timing
	if attributesVal := ctx.Value(MetricAttributesKey); attributesVal != nil {
		if attrs, ok := attributesVal.(metric.MeasurementOption); ok {
			b.readLatencyHistogram.Record(ctx, durationUs, attrs)
		}
	}

	return nil
}

func (b *ReadNarrowResultSetBenchmark) decodeRow(row *spanner.Row) {
	var randomInt64_1 int64
	var randomInt64_2 int64

	row.Column(0, &randomInt64_1)
	row.Column(1, &randomInt64_2)
}

File: java/src/main/java/com/google/cloud/spanner/benchmark/ReadNarrowResultSetBenchmark.java
package com.google.cloud.spanner.benchmark;

import com.google.cloud.spanner.DatabaseClient;
import com.google.cloud.spanner.ResultSet;
import com.google.cloud.spanner.Statement;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.LongHistogram;
import java.time.Duration;
import javax.annotation.Nonnull;

public class ReadNarrowResultSetBenchmark extends AbstractBenchmark {

  private static final String SQL =
      "SELECT\n"
          + "  FARM_FINGERPRINT(GENERATE_UUID()) AS random_int64_1,\n"
          + "  FARM_FINGERPRINT(GENERATE_UUID()) AS random_int64_2\n"
          + "FROM UNNEST(GENERATE_ARRAY(1, @num_rows)) AS n";

  private final Statement statement;
  private final Attributes customAttributes;

  public ReadNarrowResultSetBenchmark(
      DatabaseClient client,
      LongHistogram latencyHistogram,
      LongCounter operationCounter,
      LongCounter errorCounter,
      LongHistogram memoryUsageHistogram,
      DoubleHistogram cpuUtilizationHistogram,
      String resourceProbeInterval,
      String tableName,
      long minId,
      long maxId,
      double tps,
      int threads,
      Duration duration,
      boolean forAlerting,
      String benchmarkName,
      long numRows,
      LoadType loadType,
      Duration cycleDuration,
      double peakFactor,
      double burstFactor,
      double burstDuration,
      double burstFraction) {
    super(
        client,
        latencyHistogram,
        operationCounter,
        errorCounter,
        memoryUsageHistogram,
        cpuUtilizationHistogram,
        resourceProbeInterval,
        tableName,
        minId,
        maxId,
        tps,
        threads,
        duration,
        forAlerting,
        benchmarkName,
        loadType,
        cycleDuration,
        peakFactor,
        burstFactor,
        burstDuration,
        burstFraction,
        false);
    this.customAttributes = super.getAttributes().toBuilder().put("num_rows", numRows).build();
    this.statement = Statement.newBuilder(SQL).bind("num_rows").to(numRows).build();
  }

  @Override
  @Nonnull
  protected Attributes getAttributes() {
    return this.customAttributes;
  }

  // INTENTIONAL: Do not change shouldMeasureEntireMethod to return true.
  // We intentionally exclude the initial query execution and the first row fetch
  // to measure purely the iteration and decoding latency of the remaining rows.
  @Override
  protected boolean shouldMeasureEntireMethod() {
    return false;
  }

  @Override
  protected void executeOperation() throws Exception {
    try (ResultSet resultSet = client.singleUse().executeQuery(statement)) {
      if (resultSet.next()) {
        // Decode first row fully
        int dummy = decodeRow(resultSet);

        // Measure iteration of remaining rows
        long startTime = System.nanoTime();
        while (resultSet.next()) {
          dummy += decodeRow(resultSet);
        }
        long endTime = System.nanoTime();
        long latencyNs = endTime - startTime;
        long latencyUs = latencyNs / 1000;
        latencyHistogram.record(latencyUs, getAttributes());

        // Use dummy to prevent optimization
        if (dummy == 0xDEADBEEF) {
          System.out.println("This should rarely happen: " + dummy);
        }
      }
    }
  }

  private int decodeRow(ResultSet resultSet) {
    int h = 0;
    h = 31 * h + Long.hashCode(resultSet.getLong(0));
    h = 31 * h + Long.hashCode(resultSet.getLong(1));
    return h;
  }

  @Override
  protected String getBenchmarkName() {
    return "Read Narrow Result Set Benchmark";
  }

  @Override
  protected String getBenchmarkType() {
    return "read-narrow-result-set";
  }
}
